hugr-py/src/hugr/hugr/render.py

```diff
@@ -0,0 +1,280 @@
+        return f"{n.idx}:{self._OUTPUT_PREFIX}None"
+
+    def _viz_node(self, node: Node, hugr: Hugr, graph: Digraph) -> None:
+        """Render a (possibly nested) node to a graphviz graph."""
+        # TODO: Port the CFG special-case rendering from guppy, and use it here
+        # when a node is a CFG node.
+        # See https://github.com/CQCL/guppylang/blob/7d5106cd59ad452046d0dfffd10eea9d9b617431/guppylang/hugr_builder/visualise.py#L250
+
+        meta = hugr[node].metadata
+        if len(meta) > 0:
+            data = "<BR/><BR/>" + "<BR/>".join(
+                f"{key}: {value}" for key, value in meta.items()
+            )
+        else:
+            data = ""
+
+        in_ports = [str(i) for i in range(hugr.num_in_ports(node))]
+        out_ports = [str(i) for i in range(hugr.num_out_ports(node))]
+        inputs_row = (
+            self._html_ports(in_ports, self._INPUT_PREFIX) if len(in_ports) > 0 else ""
+        )
+        outputs_row = (
+            self._html_ports(out_ports, self._OUTPUT_PREFIX)
+            if len(out_ports) > 0
+            else ""
+        )
+
+        op = hugr[node].op
+
+        if hugr.children(node):
+            with graph.subgraph(name=f"cluster{node.idx}") as sub:
+                for child in hugr.children(node):
+                    self._viz_node(child, hugr, sub)
+                html_label = self._format_html_label(
+                    node_back_color=self.palette.edge,
+                    node_label=str(op),
+                    node_data=data,
+                    border_colour=self.palette.port_border,
+                    inputs_row=inputs_row,
+                    outputs_row=outputs_row,
+                )
+                sub.node(f"{node.idx}", shape="plain", label=f"<{html_label}>")
+                sub.attr(label="", margin="10", color=self.palette.edge)
+        else:
+            html_label = self._format_html_label(
+                node_back_color=self.palette.node,
+                node_label=str(op),
+                node_data=data,
+                inputs_row=inputs_row,
+                outputs_row=outputs_row,
+                border_colour=self.palette.background,
+            )
+            graph.node(f"{node.idx}", label=f"<{html_label}>", shape="plain")
+
+    def _viz_link(
+        self, src_port: OutPort, tgt_port: InPort, kind: Kind, graph: Digraph
+    ) -> None:
+        edge_attr = {
+            "penwidth": "1.5",
+            "arrowhead": "none",
+            "arrowsize": "1.0",
+            "fontname": self._FONTFACE,
+            "fontsize": "9",
+            "fontcolor": "black",
+        }
+
+        label = ""
+        match kind:
+            case ValueKind(ty):
+                label = str(ty)
+                color = self.palette.edge
+            case OrderKind():
+                color = self.palette.dark
+            case ConstKind():
+                color = self.palette.const
+            case _:
```

Comment:
Maybe explicitly match against `CFKind` and add a fallthrough case with `assert_never`. Then mypy will enforce exhaustiveness